ai-catalog/samples/gemini-live-todo/src/main/java/com/android/ai/samples/geminilivetodo/ui/TodoScreenViewModel.kt

```diff
@@ -0,0 +1,244 @@
+    
+                **Interaction Steps:**
+                **Get the task id to remove or toggle a task:** If you need to remove or check/uncheck a task,
+                    you'll need to retrieve the list of items in the list first to get the task id. Don't share 
+                    the id with the user, just identify the id of the task mentioned and directly pass this id to the 
+                    tool.
+          
+                **Never share the id with the user:** you don't need to share the id with the user. It is 
+                    just here to help you perform the check/uncheck and remove operations to the list.
+    
+                **If Unsure:** If you can't determine the update from the request, politely ask the user to rephrase or try something else.
+                    """.trimIndent(),
+                )
+            }
+
+            val addTodo = FunctionDeclaration(
+                "addTodo",
+                "Add a task to the todo list",
+                mapOf("taskDescription" to Schema.string("A succinct string describing the task")),
+            )
+
+            val removeTodo = FunctionDeclaration(
+                "removeTodo",
+                "Remove a task from the todo list",
+                mapOf("todoId" to Schema.string("The id of the task to remove from the todo list")),
+            )
+
+            val toggleTodoStatus = FunctionDeclaration(
+                "toggleTodoStatus",
+                "Change the status of the task",
+                mapOf("todoId" to Schema.string("The id of the task to remove from the todo list")),
+            )
+
+            val getTodoList = FunctionDeclaration(
+                "getTodoList",
+                "Get the list of all the tasks in the todo list",
+                emptyMap(),
+            )
+
+            val generativeModel = Firebase.ai(backend = GenerativeBackend.vertexAI()).liveModel(
+                "gemini-2.0-flash-live-preview-04-09",
+                generationConfig = liveGenerationConfig,
+                systemInstruction = systemInstruction,
+                tools = listOf(
+                    Tool.functionDeclarations(
+                        listOf(getTodoList, addTodo, removeTodo, toggleTodoStatus),
+                    ),
+                ),
+            )
+
+            session = generativeModel.connect()
+            Log.d(TAG, "Gemini Live session initialized")
+            _uiState.update {
+                if (it is TodoScreenUiState.Success) {
+                    it.copy(isLiveSessionReady = true)
+                } else {
+                    it
+                }
+            }
+        }
+    }
+
+    private fun handleFunctionCall(functionCall: FunctionCallPart): FunctionResponsePart {
+        return when (functionCall.name) {
+            "getTodoList" -> {
+                val todoList = todoRepository.getTodoList().reversed()
+                val response = JsonObject(
```

Comment:
For these and the below examples, it is probably cleaner to use the json dsl like follows. You should also wrap it up in a function because it is repeated a lot

```
fun functionResponse(success: Boolean, message: String) = buildJsonObject {
   put("success", message.success)
   put("message", message.message)
}
```